### PR TITLE
feature(@desktop/chat): support jumping to search result message that is not currently loaded in memory

### PR DESCRIPTION
**Pull request by saledjenic**
This branch is branched off [feature/issue-2912](https://github.com/status-im/status-desktop/tree/feature/issue-2912) which PR is [here](https://github.com/status-im/status-desktop/pull/3008). It requires [status-go PR](https://github.com/status-im/status-go/pull/2282) to be approved and added first.

Changes made here (which were necessary in order to have good base to develop requested feature):
- duplicated methods are removed
- loading messages either on start (initial loading) or loading more messages requested by scrolling in the chat view is now done in separate thread (asynchronous) instead as it was earlier, done in the main thread
- new file is added for storing async tasks for chat part only
- ChatModel (from status/chat.nim) is QObject instance now, so it may handle async job in a slot
- a job requested from the view is done in separate thread and view is notified about the changes using signal/slot mechanism

This is not refactoring, but minimal update to have clear situation for further changes.

I split that in two commits:
- Necessary changes [here](https://github.com/status-im/status-desktop/commit/c960c2343c56cf9b2bf2fa4768c3a9777f393798)
- Feature development (jumping to search result message that is not currently loaded) [here](https://github.com/status-im/status-desktop/commit/1a04b3315febefc8c69919c901bd4426c6a9a648) (you may notice how easily is added this feature in this commit)

_New known issue_:
- app crash when you click on (try to add) any public chat

_Reason for having this new issue_:
- That's happening cause I removed loading chat messages from the main thread and do that asynchronous in a separate thread. Seems we have a lot of tight coupled things between nim and qml parts, what caused this crash. Now, loading initial messages for every new channel (in event_handling.nim calls ```self.status.chat.loadInitialMessagesForChannel(channel.chat.id)``` ) is done asynchronous and that's corrupted order of execution which was applied before by loading initial chat messages in the main thread.

_Solution_:
1. That may be solved (quick and easy but not recommended way) simply by loading initial messages in the main thread
2. Searching for that bug and fixing it appropriately (recommended way), as I briefly see that in that case we are calling setActive channel multiple times and emit notifying signals while the property is being changed what causes crash in qml part

All in all whatever solution we decide to proceed with, changes made here follows some kind of architectural flow and I suggest to fix that in a separate ticket, not in this one, but you guys make a final decision.

### CODE REVIEW

File: src/app/chat/views/messages.nim [hunk 1]
```diff
@@ -100,11 +57,34 @@ QtObject:
...
-    result.observedMessageItem = newMessageItem(status)
+    result.searchResultMessageModel = newMessageListProxyModel(status)
     result.unreadDirectMessagesAndMentionsCount = 0
+    result.loadingHistoryMessages = false
     result.setup
 
-  # proc getMessageListIndexById(self: MessageView, id: string): int
+  #################################################
+  # Forward declaration section
+  #################################################
```
Comment (alaibe) [inline src/app/chat/views/messages.nim:67]: I am not sure if there is consensus here on those type of comment but if there isn't that will quickly become unmaintained and source of confusion

File: src/status/chat.nim [hunk 2]
```diff
@@ -73,550 +62,697 @@ type
     notificationTypes*: seq[ActivityCenterNotificationType]
 
 include chat/utils
+include chat/async_tasks
+
+QtObject:
```
Comment (alaibe) [inline src/status/chat.nim:67]: I am not totally sure here but I don't think src/status is supposed to be couple with QT. I might be wrong but I think it should only be from src/app

File: src/app/chat/views/messages.nim (hunk 1 above)
Comment (saledjenic) [reply src/app/chat/views/messages.nim:67]: @alaibe we should agreed on this, if that's an issue I can remove it, it's not a problem at all, but I found this kind of headers like "Forward declaration section", "Property section" very useful cause that increases readability and avoids some duplications I would say,

File: src/status/chat.nim (hunk 2 above)
Comment (saledjenic) [reply src/status/chat.nim:67]: @alaibe I must say that I cannot agree on this with you. :) 
As you noticed nim source code is divided in two parts "app" which should be used for view stuff and "status" where should be placed app's logic. We're developing a Qt app, that means we're using Qt in both parts, in view part (where usually models like QAbstractListModel and other should be used) and logic part where we must inherit QObject if we want to work with signal/slot mechanism. So it's necessary on both sides.

File: src/app/chat/views/messages.nim (hunk 1 above)
Comment (alaibe) [reply src/app/chat/views/messages.nim:67]: I don't have personal preference on that, for or against, I am happy to follow. So for me not an issue but as soon as we all follow this guideline I am good!

File: src/status/chat.nim (hunk 2 above)
Comment (alaibe) [reply src/status/chat.nim:67]: As I said, I wasn't totally sure it was more of an impression from looking at the code itself. Maybe @jrainville @richard-ramos  could help here?

**Review by jrainville — COMMENTED** [review]
Changes look very good.

As for the new bug/crash, I think we need to fix it before merging this, because it seems like a big showstopper.

You proposed two options. Would the longer, better option be doable on top of this branch in another PR in a week, as to reduce conflicts?

If not, I'd say we could use the faster option of putting back the synchonous functions, while still keeping your other code, and just open another issue to remove the sync functions and just use your async code, but with the fixes to the crash.

What do you think?

File: src/app/chat/views/messages.nim (hunk 1 above)
Comment (saledjenic) [reply src/app/chat/views/messages.nim:67]: @jrainville when I say "Properties" I referred to QtProperty. Yes, they are simple functions, but generally speaking they are usually short and do a single action (it's not recommended to have WRITE property which will do a lot of other calls) they are created to update a view (qml) quickly. 

In c++ you may use Q_PROPERTY macro in header files and when you look the class header all properties are on the same place and you may get an easy idea what was an intention there and what is exposed to qml as property. We don't have header files here and just wanted to wanted to emphasize part where we're stating Qt's properties, that's just a matter of style (I don't like to search through the whole file for a properties), but as you decide.

Comment (saledjenic) [reply src/app/chat/views/messages.nim:67]: @jrainville when I say "Properties" I referred to QtProperty. Yes, they are simple functions, but generally speaking they are usually short and do a single action (it's not recommended to have WRITE property which will do a lot of other calls) they are created to update a view (qml) quickly.

In c++ you may use Q_PROPERTY macro in header files and when you look the class header all properties are on the same place and you may get an easy idea what was an intention there and what is exposed to qml as property. We don't have header files here and just wanted to wanted to emphasize part where we're stating Qt's properties, that's just a matter of style (I don't like to search through the whole file for a properties), but as you decide.

File: src/app/chat/views/messages.nim [hunk 3]
```diff
@@ -513,9 +447,29 @@ QtObject:
...
+  proc onSearchMessagesLoaded*(self: MessageView, messages: seq[Message]) =
+    self.searchResultMessageModel.setFilteredMessages(messages)
+
+  proc isMessageDisplayed*(self: MessageView, messageId: string): bool {.slot.} =
+    let chatId = self.channelView.activeChannel.id
+    var message = self.messageList[chatId].getMessageById(messageId)
+
+    return message.id != ""
+
+  proc loadMessagesTillMessageWithIdIsLoaded*(self: MessageView, messageId: string) {.slot.} =
```
[the remark this answers is not in the document]
Comment (saledjenic) [reply src/app/chat/views/messages.nim:461]: @jrainville I agree name is not pretty but describes what the procedure does. :) Any suggestion for the name?

[the remark this answers is not in the document]
Comment (jrainville) [reply src/app/chat/views/messages.nim:461]: Just need to change `Till` to `Until` so it's formal English

File: src/app/chat/views/messages.nim [hunk 4]
```diff
@@ -513,9 +447,29 @@ QtObject:
...
-      slot: "onSearchMessages"
-    )
+    self.status.chat.asyncSearchMessages(chatId, searchTerm, false)
 
-    self.status.chat.asyncSearchMessages(slot, chatId, searchTerm, false)
\ No newline at end of file
+  proc onSearchMessagesLoaded*(self: MessageView, messages: seq[Message]) =
+    self.searchResultMessageModel.setFilteredMessages(messages)
+
+  proc isMessageDisplayed*(self: MessageView, messageId: string): bool {.slot.} =
```
[the remark this answers is not in the document]
Comment (saledjenic) [reply src/app/chat/views/messages.nim:455]: @jrainville actually meant on displayed messages. A model is (should be) actually what is displayed (what user sees). And in general ```isMessageLoaded``` means that it's loaded (it's somewhere in memory), but doesn't imply it's displayed and for this ticket particularly you should point a user to appropriate message which must be displayed.

**Comment by saledjenic** [comment]
> You proposed two options. Would the longer, better option be doable on top of this branch in another PR in a week, as to reduce conflicts?
> 
> If not, I'd say we could use the faster option of putting back the synchonous functions, while still keeping your other code, and just open another issue to remove the sync functions and just use your async code, but with the fixes to the crash.
> 
> What do you think?

@jrainville yes you get it well, and I guess that your second option will be better and easier to do. The only change which I need to do to make it work without a crash is calling loadInitialMessagesForChannel synchronous. Searching messages and loading more messages will still remain asynchronous. So all current code remain only I need to add one more part for fetching messages from the main thread.

And after that is approved and merged into the master we may continue with those improvements for loading initial messages asynchronous.

**Comment by pinebit** [comment]
Fixes #3064
I tested this branch and among other things this fixed my issue I was working on (#3064)
Thanks :)

**Review by pinebit — APPROVED** [review]

**Comment by saledjenic** [comment]
@jrainville I fixed the issue related to the crash which started to happen after my initial changes for this feature. Now everything works fine.

While I was working on this I noticed one very hidden issue which was causing app crash. It's happening when you "Leave chat", not always but may be faced very often. The thing is that "messages" exposed to qml were deleted from ChatsView and signals beginRemoveRows and endRemoveRows were emitted for CatsView abstract list model instead for messageView abstract list model. That results in an app crash in some moments (not always reproducible, but often).

Fix for that only is in [this commit](https://github.com/status-im/status-desktop/commit/3fb8b766779f10befc3d44c7c46fd525e1311941)

@iurimatias regardless to this feature, you may consider picking that commit for you next release as it is something we will need to avoid potential crash.

**Pull request merged**